spark/src/main/scala/com/gu/tableversions/spark/VersionedFileSystem.scala

```diff
@@ -0,0 +1,159 @@
+package com.gu.tableversions.spark
+
+import java.net.URI
+import java.util.Objects
+
+import com.gu.tableversions.core.Partition
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.fs._
+import org.apache.hadoop.fs.permission.FsPermission
+import org.apache.hadoop.util.Progressable
+
+abstract class ProxyFileSystem extends FileSystem {
+
+  protected var pathMapper: PathMapper = _
+
+  def setPathMapper(mapper: PathMapper): Unit =
+    pathMapper = mapper
+}
+
+class VersionedFileSystem extends ProxyFileSystem {
+
+  private var baseURI: URI = _
+  private var baseFS: FileSystem = _
+  override def initialize(name: URI, conf: Configuration): Unit = {
+    val cacheDisabled = conf.getBoolean("fs.versioned.impl.disable.cache", false)
+    val baseFSScheme = conf.get("fs.versioned.baseFS")
+    val version = conf.get("fs.versioned.version")
+
+    require(Objects.nonNull(baseFSScheme), "fs.versioned.baseFS not set in configuration")
+    require(Objects.nonNull(version), "fs.versioned.version not set in configuration")
+    require(cacheDisabled, "fs.versioned.impl.disable.cache not set to true in configuration")
+
+    import scala.collection.JavaConverters._
+
+    val partitionMappings = conf.iterator.asScala
+      .filter(_.getKey.startsWith(VersionedFileSystem.keyPrefix))
+      .map(_.getValue)
+      .toList
+      .map { v =>
+        (v, v + "/" + version)
+      }
+      .toMap
+
+    setPathMapper(new VersionedPathMapper(baseFSScheme, partitionMappings))
+
+    // When initialising the versioned filesystem, we need to extract the
+    // URI for the base filesystem, which is accessible under the scheme
+    // specific part of the URI.
+    baseURI = new URI(baseFSScheme, null, name.getSchemeSpecificPart, null, null)
+    baseFS = FileSystem.get(baseURI, conf)
+    baseFS.initialize(baseURI, conf)
+  }
+
+  override def getUri: URI =
+    new URI(s"${VersionedFileSystem.scheme}:${baseURI.getSchemeSpecificPart}")
+
+  override def open(f: Path, bufferSize: Int): FSDataInputStream =
+    baseFS.open(pathMapper.forUnderlying(f), bufferSize)
+
+  override def create(
+      f: Path,
+      permission: FsPermission,
+      overwrite: Boolean,
+      bufferSize: Int,
+      replication: Short,
+      blockSize: Long,
+      progress: Progressable): FSDataOutputStream =
+    baseFS.create(pathMapper.forUnderlying(f), permission, overwrite, bufferSize, replication, blockSize, progress)
+
+  override def append(f: Path, bufferSize: Int, progress: Progressable): FSDataOutputStream =
+    baseFS.append(pathMapper.forUnderlying(f), bufferSize, progress)
+
+  override def rename(src: Path, dst: Path): Boolean =
+    baseFS.rename(pathMapper.forUnderlying(src), pathMapper.forUnderlying(dst))
+
+  override def delete(f: Path, recursive: Boolean): Boolean =
+    baseFS.delete(pathMapper.forUnderlying(f), recursive)
+
+  private def toggleFileStatus(fileStatus: FileStatus, isParam: Boolean): FileStatus = {
```

Comment:
How about replacing the `isParam` argument with an argument `convert: Path => Path` instead? Then the caller can just pass in `pathMapper.fromUnderlying` or `pathMapper.forUnderlying`.

(And change the name from `toggle...` to `convert...`  or `map...`?)